generate/download.go

```diff
@@ -45,10 +48,21 @@ func CheckOrDownloadProtoc(path, version string) (string, error) {
 		dstPath = filepath.Join(cacheDir, fmt.Sprintf("protoc-%s", version))
 	}
 
-	// Check the cache path to see if it has been previously
-	// downloaded by gunk.
+	// First, grab a lock separate from the destination file. The
+	// destination file is a binary we'll want to execute, so using it
+	// directly as the lock can lead to "text file busy" errors.
+	unlock, err := lockedfile.MutexAt(dstPath + ".lock").Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer unlock()
+	// We are the only goroutine with access to dstPath. Check if it already
+	// exists. Using lockedfile.OpenFile allows us to do an atomic write
+	// when it doesn't exist yet.
 	dstFile, err := lockedfile.OpenFile(dstPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0775)
```

Comment:
my understanding is we already using a lock file mutex above, so lockedfile.OpenFile could be replaced with os.OpenFile without any ill effect. Is this correct?

Reply:
Not exactly, because we still want the file write to be atomic. This is what the comment just above explains. For example, if the download fails, or if the user hits ^C while we are downloading or writing the file, or if the power shuts off mid-process, we don't want to leave a broken protoc binary.

Reply:
right, sorry I didn't read the comment carefully.